### PR TITLE
Feature: Add "Used on Current Site" List of Masks to in-page menu

### PR DESCRIPTION
**Pull request by maxxcrawford**
## Summary

[MPP-1638](https://mozilla-hub.atlassian.net/browse/MPP-1638) 

Blocked (Cannot submit to AMO until) by: 
- https://github.com/mozilla/fx-private-relay/pull/1968
- https://github.com/mozilla/fx-private-relay/pull/1969

This PR adds additional functionality to the in-page menu. Now, when users fill in email inputs using the add-on, we will remember that and service it the next time they're on the site. 

## Caveats

- ~Local storage only (not syncing with the server) has not been tested/is not currently supported~
- ~Website limitation:  `domainaddresses` does not yet support the `used_on` field~
- This functionality does not apply to context-menu/alt-clicking
  - Draft started https://github.com/mozilla/fx-private-relay-add-on/pull/342

### Free account: 

- Start with an empty account (with sync enabled)
- Create a few masks from the dashboard page
  - _Optional: Set a label_
- Go to a website with an email input form
- Click on the Relay icon
- **Expected:** The in-page menu should open up with the masks in the "Select from your current email masks" list
  - _Note: If you set a label, it should show the label in this list, rather than the mask address._
- Click on a mask
- **Expected:** The mask address should be auto-filled into the email input
- Click on the Relay icon again
- **Expected:** The in-page menu should open up with that selected mask now in the "Previously used on this website" list 
  - _Note: If you created multiple masks, there will now be two lists: "Select from your current email masks" and "Previously used on this website"_
- Click on "Generate new mask" button
- **Expected:** A mask address should be created and then auto-filled into the email input
- Click on the Relay icon again
- **Expected:** There should be two items in the "Previously used on this website" list 
  - _Note: The newest mask address should have a label of the current website you're on._

### Premium Account

- Start with a premium, empty account (with sync enabled)
- Create a few masks from the dashboard page
  - _Note: If you create a custom mask (with a registered subdomain), this will NOT work until https://github.com/mozilla/fx-private-relay/pull/1968 is merged. You can still re-use these with the add-on, but it will not save/store site usage._
- Go to a website with an email input form
- **Expected:** The in-page menu should open up with the masks from a single (unlabeled) list. 
  - _Optional: If you created over five masks, you should have a search field to search that visible list of masks._
- Click on a mask
- **Expected:** The mask address should be auto-filled into the email input
- Click on the Relay icon again
- **Expected:** There are now TWO lists (with navigation buttons at the top of the menu): From this website / All email masks. The email mask you selected previously is in "From this website" list.
  - _Note: The "All email masks" will list every mask created for the user, including ones from on the "From this website" list_
- Click on "Generate new mask" button
- **Expected:** A mask address should be created and then auto-filled into the email input
- Click on the Relay icon again
- **Expected:** There should be two items in the "From this website" list 
  - _Note: The newest mask address should have a label of the current website you're on._
- Go back to the dashboard and create enough masks to have at least six.
- Go back to the same previous website
- Click on the Relay icon
- **Expected:** There should still be two lists: From this website / All email masks 
- Click on "All email masks" label to switch lists
- **Expected:** There should a search bar that appears (and it should filter as expected too!) 

## Screenshots

Free account with know mask association with current site: 

image

Premium account with know mask association with current site: 
image

image

### CODE REVIEW

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:105]: nit (non-blocking) micro-optimization?: Is `String.split()` and then Array.includes()` faster than `String.includes()`?

If `String.includes()` is as-fast or faster, we could change this to check `domainList.includes(currentDomain)` and `return` before calling `split(",")`

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:213]: suggestion (non-blocking) Is is still considered faster to use `createDocumentFragment` instead of `createElement`?

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:293]: praise: From all of us less fluent in JS, TY for making this a nicely-named function instead of a terse, dense, and cryptic 1-liner. 😄

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:306]: suggestion (non-blocking): similar to above, do we have to `split()` the string into an array and then call `includes()`, or can we just call `includes()` directly on the string?

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:325]: suggestion (non-blocking): again, maybe we can call `includes()` directly on the string here?

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:375]: `TODONE`! This is merged now, so can we update this comment and line?

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:387]: This `TODO` is also `TODONE`!

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:426]: TIL `await Promise` evaluates before `Object?`.

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:503]: 🦅 👀 
```suggestion
      // If so, we'll end up building two discrete mask lists:  
```

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:545]: `TODONE`!

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:786]: nit (non-blocking): this function is doing more than just focusing on the search input - it's also updating the visible/total text. Maybe a better function name to match what it's doing?

**Review by groovecoder — COMMENTED** [review]
Lots of non-blocking code comments, but I haven't been able to spot-check yet, so not approving yet.

File: src/js/background/background.js
Comment (maxxcrawford) [reply src/js/background/background.js]: Good call. I updated a few other variable names too! 

Revised in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js]: 🙈 I really need comment spell-check. 

Revised in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js]: Renamed this function to `serverStoragePrefInLocalStorage` to make it a bit more readable. Revised in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js]: It totally works. Great call out and opportunity for optimization. Revised in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js:217]: Replaced for the `li` `createElement`. The other element (`button`) is still using the other method, but is being injected into the fragment. Should be good to go. 😎 

Revised in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js]: Revised in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js]: Revised in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js]: Updated to use the L10N string. Revised in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js]: Removed in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js]: Revised in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js]: Removed in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (maxxcrawford) [reply src/js/other-websites/inpage_menu.js]: Renamed function to `initResultsCountAndFocusOnInput()`. Revised in d5e6221fe717faa6809d549a9e0b85259f67a063

File: src/js/other-websites/inpage_menu.js
Comment (groovecoder) [inline src/js/other-websites/inpage_menu.js:108]: such concise! so ternary!

**Review by groovecoder — APPROVED** [review]
Code and spot-checks look good. Except the address is still coming up hard-coded to @mozmail.com instead of the `full_address` field from the API return value. (I noticed it from the in-page pop-up and also from the context menu)

Approving after that's fixed. Don't forget to build the QA add-on for stage server. ;)

**Comment by maxxcrawford** [comment]
@Vinnl After discussing in stand-up, I am reassigning this PR review to @codemist. Feel free to review after it merges for any serious issues. There is some time between this merging and the add-on being submitted so you still have time!

**Review by codemist — COMMENTED** [review]
Overall, this feature works well and it's super snappy!

I did however encounter two main bugs.

1) For nonpremium, I was only able to see the masks generated on the site under the "Previously used on this website" tab, it's missing the pre-existing masks that were used on the site.
https://user-images.githubusercontent.com/13066134/169382584-fcc69484-31eb-4eaa-bf21-f51e52ea3613.mov

2) When switching accounts (free to premium) the masks of the free account did not update to masks of the premium account even after refreshing, logging out and then back in, and opening the "Manage" CTA. 
https://user-images.githubusercontent.com/13066134/169383181-3de0f507-3895-4fc4-a4ca-069fa7249332.mov

Third one's just a design inconsistency that I noted:
3) Seems like the UI for a mask with a label varies between premium and nonpremium versions. Not sure if that's intentional, but maybe we should keep it consistent for accessibility (see [Recognition > Recall](https://www.nngroup.com/articles/recognition-and-recall/#:~:text=The%20first%20process%20is%20recognition,of%20related%20details%20from%20memory.) heuristic) 
image

File: src/css/in-page.css
Comment (codemist) [inline src/css/in-page.css:371]: Recommend using `overflow-y` instead so it only activates vertical scrolling.

Comment (codemist) [reply src/css/in-page.css:371]: UI tweak nit: I also added `scrollbar-width: thin;` and set the footer to a fixed bottom sticky position to remove the odd spacing surrounding the mask list.

Before:
image

After:

https://user-images.githubusercontent.com/13066134/169396746-abe2b846-bfba-4d5a-9a24-a35e5a417ce0.mov

**Review by codemist — APPROVED** [review]

File: src/css/in-page.css
Comment (Vinnl) [inline src/css/in-page.css:26]: Oh yeah this is a much better place than on line 28 :P

File: src/css/in-page.css
Comment (Vinnl) [inline src/css/in-page.css:255]: Could you add a comment explaining which element this z-index is intending to overlap? To help future contributors tell if it can be removed/upped.

File: src/css/in-page.css
Comment (Vinnl) [inline src/css/in-page.css:325]: Likewise: z-index comment?

File: src/js/background/background.js
Comment (Vinnl) [inline src/js/background/background.js:26]: Nit, but I'd expect names starting with a verb (`get*`) to be functions.

File: src/js/background/background.js
Comment (Vinnl) [inline src/js/background/background.js:85]: Wasn't there a function somewhere that did all this?

**Review by Vinnl — COMMENTED** [review]
Given that there are already two reviews (and that I wouldn't have been able to form a mental model of how it all folds together anyway), I just did a quick scan through to see if anything jumped out and added some suggestions you could consider (or ignore) :)

File: src/js/background/background.js
Comment (maxxcrawford) [reply src/js/background/background.js:75]: Yes! Replaced with [`createNewHeadersObject()` function](https://github.com/mozilla/fx-private-relay-add-on/blob/main/src/js/background/background.js#L170-L186). Replaced in ﻿﻿80402c57eaa99b1b1617691f015a18f1ed6fda02

File: src/js/background/background.js
Comment (maxxcrawford) [reply src/js/background/background.js]: Renamed all of the API variables to follow this pattern: 
- `relayApiUrlRelayAddresses`
- `relayApiUrlRelayAddressId`
- `relayApiUrlDomainAddresses`

Revised in 80402c57eaa99b1b1617691f015a18f1ed6fda02

File: src/css/in-page.css
Comment (maxxcrawford) [reply src/css/in-page.css:329]: Comment added in 80402c57eaa99b1b1617691f015a18f1ed6fda02

File: src/css/in-page.css
Comment (maxxcrawford) [reply src/css/in-page.css:268]: Comment added in 80402c57eaa99b1b1617691f015a18f1ed6fda02

File: src/css/in-page.css
Comment (maxxcrawford) [reply src/css/in-page.css:256]: Comment added in 80402c57eaa99b1b1617691f015a18f1ed6fda02

File: src/css/in-page.css
Comment (maxxcrawford) [reply src/css/in-page.css]: Whoa! TIL about `scrollbar-width: thin`. Nice fix! 

Added in ac4214cc685a936df860006fe1b376b448e81ccd

**Pull request merged**
